cifar10-multiannotator/utils/model_training.py

```diff
@@ -0,0 +1,272 @@
+
+    time_limit : int, default=7200
+      Passed into AutoGluon's `ImagePredictor().fit()` method.
+
+    random_state : int, default=123
+      Passed into AutoGluon's `ImagePredictor().fit()` method.
+
+    Returns
+    -------
+    None
+
+    """
+
+    # stratified K-folds
+    skf = StratifiedKFold(n_splits=n_splits, shuffle=False)
+    skf_splits = [
+        [train_index, test_index]
+        for train_index, test_index in skf.split(X=dataset, y=dataset.label)
+    ]
+
+    # run cross-validation
+    for split_num, split in enumerate(skf_splits):
+
+        print("----")
+        print(f"Running Cross-Validation on Split: {split_num}")
+
+        # TODO: add logic to skip if results from model/fold already exists from a previous run
+
+        # split from stratified K-folds
+        train_index, test_index = split
+
+        # init model
+        predictor = ImagePredictor(verbosity=0)
+
+        # train model on train indices in this split
+        predictor.fit(
+            train_data=dataset.iloc[train_index],
+            ngpus_per_trial=ngpus_per_trial,
+            hyperparameters=model_params,
+            time_limit=time_limit,
+            random_state=random_state,
+        )
+
+        # predict on test indices in this split
+
+        # predicted probabilities for test split
+        pred_probs = predictor.predict_proba(
+            data=dataset.iloc[test_index], as_pandas=False
+        )
+
+        # predicted features (aka embeddings) for test split
+        # why does autogluon predict_feature return array of array for the features?
+        # need to use stack to convert to 2d array (https://stackoverflow.com/questions/50971123/converty-numpy-array-of-arrays-to-2d-array)
+        pred_features = np.stack(
+            predictor.predict_feature(data=dataset.iloc[test_index], as_pandas=False)[
+                :, 0
+            ]
+        )
+
+        # save output of model + split in pickle file
+
+        out_subfolder = f"{out_folder}split_{split_num}/"
+
+        try:
+            os.makedirs(out_subfolder, exist_ok=False)
+        except OSError:
+            print(f"Folder {out_subfolder} already exists!")
+        finally:
+
+            # save to pickle files
+
+            get_pickle_file_name = (
+                lambda object_name: f"{out_subfolder}_{object_name}_split_{split_num}"
+            )
+
+            _save_to_pickle(pred_probs, get_pickle_file_name("test_pred_probs"))
+            _save_to_pickle(pred_features, get_pickle_file_name("test_pred_features"))
+            _save_to_pickle(
+                dataset.iloc[test_index].label.values,
+                get_pickle_file_name("test_labels"),
+            )
+            _save_to_pickle(
+                dataset.iloc[test_index].image.values,
+                get_pickle_file_name("test_image_files"),
+            )
+            _save_to_pickle(test_index, get_pickle_file_name("test_indices"))
```

Comment:
It's probably not worth re-running for this particular example, but just FYI in the future we won't want to save numeric data in pickle:
https://stackoverflow.com/questions/52444921/save-numpy-array-using-pickle

It's better to use: `np.save(), np.load()` or `hdf5, h5py`